File: c/20.c
#include "common.h"

#define MAX_GRID_SIZE 512
int grid[MAX_GRID_SIZE][MAX_GRID_SIZE];
int buffer[MAX_GRID_SIZE][MAX_GRID_SIZE];
int grid_size;
int mapping[1024] = {0};
int default_val = 0;

void dump_grid() {
  for (int i = 0; i < grid_size; i++) {
    for (int j = 0; j < grid_size; j++)
      printf("%c ", grid[i][j] ? '#' : '.');
    printf("\n");
  }
}

void parse() {
  FILE *f = fopen(INFILE("20.txt"), "r");
  char buf[2048];

  GETLINE(buf, f);
  int i = 0;
  for (char *line = buf; *line; line++)
    mapping[i++] = (*line) == '#';

  GETLINE(buf, f); // Skip newline

  while (GETLINE(buf, f)) {
    int i = 0;
    for (char *line = buf; *line; line++)
      grid[grid_size][i++] = (*line) == '#';
    grid_size++;
  }

  fclose(f);
}

int pixel(int i, int j) {
  if (i < 0 || i >= grid_size || j < 0 || j >= grid_size)
    return default_val;
  return grid[i][j];
}

int mapping_index(int i, int j) {
  static const int NBS[9][2] = {
    {-1, -1}, {-1, 0}, {-1, 1},
    { 0, -1}, { 0, 0}, { 0, 1},
    { 1, -1}, { 1, 0}, { 1, 1}
  };
  int k = 0;
  for (int n = 0; n < 9; n++)
    k = k << 1 | pixel(i + NBS[n][0], j + NBS[n][1]);
  return k;
}

void enhance() {
  for (int i = -1; i <= grid_size; i++)
    for (int j = -1; j <= grid_size; j++)
      buffer[i+1][j+1] = mapping[mapping_index(i, j)];
  grid_size += 2;
  memcpy(grid, buffer, sizeof(grid));

  // Careful! in this case the whole infinite grid outside changes values
  // every iteration. Assumes the pair is always (0,1) or (1,0) only.
  if (mapping[0] == 1) // (and mapping[-1] == 0)
    default_val = !default_val;
}

int sum_on() {
  int sum = 0;
  for (int i = 0; i < grid_size; i++)
    for (int j = 0; j < grid_size; j++)
      sum += grid[i][j];
  return sum;
}

int main() {
  parse();

  int i = 0;
  for (; i < 2; i++) enhance();
  printf("Part 1: %d\n",sum_on());

  for (; i < 50; i++) enhance();
  printf("Part 2: %d\n",sum_on());

  return 0;
}

File: python/20.py
from common import *

with open(infile('20.txt')) as f:
    mapping, rest = f.read().replace('#', '1').replace('.', '0').split('\n\n')
mapping = [int(c) for c in mapping]
image, _, _ = parse_block_grid(rest)
default_value = 0

def neighbourhood(i, j):
    return ((i-1, j-1), (i-1,  j ), (i-1, j+1),
            ( i , j-1), ( i ,  j ), ( i , j+1),
            (i+1, j-1), (i+1,  j ), (i+1, j+1))

def enhance(grid):
    max_i, max_j = len(grid), len(grid[0])

    def find_value(ci, cj):
        k = 0
        for i, j in neighbourhood(ci, cj):
            v = default_value
            if 0 <= i < max_i and 0 <= j < max_j:
                v = grid[i][j]
            k = k << 1 | v
        return mapping[k]

    new = [
        [ find_value(i, j) for j in range(-1, max_j+1) ]
        for i in range(-1, max_i+1)
    ]

    # Careful! in this case the whole infinite grid outside changes values
    # every iteration. Assumes the pair is always (0,1) or (1,0) only.
    global default_value
    if mapping[0] == 1 and mapping[-1] == 0:
        default_value = 1 - default_value

    return new

def solve(iters):
    a = image
    for _ in range(iters):
        a = enhance(a)
    return sum(sum(row) for row in a)

print("Part 1:", solve(2))
print("Part 2:", solve(50))